### DISCUSSION TITLE
Auto-rendering other filetypes (e.g. .org)?

### DISCUSSION THREAD

**Discussion by benfried** [quarto-dev/quarto-cli#2151] [Q&A]
I'd like to use Emacs org mode to build my web site, and have quarto auto-render the .org files into .md as they are added and updated --- this could be either with pandoc or a custom script that wraps pandoc.

Any pointers on how to do this? I suppose I can always work around with a makefile, or an emacs after-save-hook, but those seem so much less elegant than having quarto know how to render the .org file.

I've looked through the documentation, and maybe I'm just not looking for the right things, but I have not figured out how to do this.

Thanks...

Ben

**Comment by cscheid** [comment]
You can use pre-render scripts to populate a quarto project with .md files from .org: https://quarto.org/docs/projects/scripts.html

**Reply by benfried** [reply]
Thanks! Looks like pre-render and post-render were missing from schema/project.yml until you put them back just a bit ago...

**Reply by cscheid** [reply]
That's right, truly a coincidence (someone else reported the bug!)

**Reply by Rahuketu86** [reply]
I am trying to address a similar issue and have a hackish solution with pre-render script for docx file format. Why hackish? Because in addition to providing a docx file I need to manually add an empty file "index.qmd"  and a separate "frontmatter.txt". 
<img width="233" alt="image" src="https://user-images.githubusercontent.com/786803/187068661-fd92cde1-7cfe-4f8b-bccd-00dc725cd318.png">

I guess there is little we can do about frontmatter.txt but I don't like creating an empty index.qmd . Reason I need to that becoz in pre-render script environment variable  'QUARTO_PROJECT_INPUT_FILES' doesn't pick up files with .docx as input files and even though the code executes it is not rendered without the presence of an index.qmd file. Is there a better way to enable files of a particular extension(.docx, org) to be assigned to the initial input files list ?

```
import os
import glob
import hashlib
import subprocess 
import contextlib

@contextlib.contextmanager
def chdir(target):
    try:
        start = os.getcwd()
        os.chdir(target)
        yield

    finally:
        os.chdir(start)


def filter_names(folder='posts', ext=".docx"):
    # return [fname  for fname in os.getenv('QUARTO_PROJECT_INPUT_FILES') if fname.endswith(ext)]
    return glob.glob(f"{folder}/**/*{ext}", recursive=True)


def checksum(fname):
    pass


def docx2qmd():
    curdir = os.getcwd()
    converted_qmd = []
    for fname in filter_names():
        dirname = os.path.dirname(os.path.abspath(fname))
        with chdir(dirname):
            standalone_fname = os.path.basename(fname)
            frontmatter = open("frontmatter.txt", 'rb').readlines()
            cmd = f"quarto pandoc --from docx --to gfm --output generated.md --columns 9999 --extract-media=. --standalone {standalone_fname}"
            out = subprocess.check_output(cmd, shell=True)

            with open("index.qmd", "ab") as index_qmd, open("frontmatter.txt", "rb") as frontmatter, open("generated.md", "rb") as generated_md:
                index_qmd.write(frontmatter.read())
                index_qmd.write(b"\n")
                index_qmd.write(generated_md.read())
            os.unlink("generated.md")
            converted_qmd.append(os.path.relpath(os.path.abspath("index.qmd"), curdir))
    return converted_qmd



def org2qmd():
    pass

def convert():
    # print(os.getenv('QUARTO_PROJECT_RENDER_ALL'))
    # print(os.getenv('QUARTO_PROJECT_OUTPUT_DIR'))
    print(os.getenv('QUARTO_PROJECT_INPUT_FILES'))
    # print(os.getenv('QUARTO_PROJECT_OUTPUT_FILES'))
    # print(list(filter_names()))
    docqmds = docx2qmd()
    print(docqmds)
    input_files = os.getenv('QUARTO_PROJECT_INPUT_FILES')+"\n"+"\n".join(docqmds)
    print(input_files)
    os.environ['QUARTO_PROJECT_INPUT_FILES'] = input_files


if __name__=='__main__':
    convert()
```

Attempts to modify the environment variable ('QUARTO_PROJECT_INPUT_FILES'  has not worked.

 My _quarto.yml is as follows 

```
project:
  type: website
  pre-render: custom2qmd.py

website:
  title: "qblog"
  navbar:
    right:
      - about.qmd
      - icon: github
        href: https://github.com/
      - icon: twitter
        href: https://twitter.com
format:
  html:
    theme: cosmo
    css: styles.cs
```s

**Comment by benfried** [comment]
It turns out that quarto doesn't watch for file types that it doesn't know about, so there was no way to trigger a pre-render script to convert org to md. Instead, I wrote an emacs after-save-hook that figures out if the org file is in a quarto content directory, and if it is, runs pandoc to convert the .org to .md; works just fine.

I did end up using a pre-render script to generate the site table of contents.

Happy to share the emacs bits or the pre-render script if there's any interest.

Ben

**Reply by Rahuketu86** [reply]
Please share both. Maybe I can optimize my solution.
